src/View/Helper/FormHelper.php

```diff
@@ -2672,4 +2675,29 @@ public function implementedEvents()
     {
         return [];
     }
+
+    public function getValuesSources()
+    {
+        return $this->_valuesSources;
+    }
+
+    public function setValuesSources($sources)
+    {
+        $this->_valuesSources = array_values(array_intersect((array)$sources, ['context', 'data', 'query']));
+
+        return $this;
+    }
+
+    public function getSourceValue($fieldname)
+    {
+        foreach ($this->getValuesSources() as $valuesSource) {
+            if ($valuesSource === 'context') {
+                return $this->_getContext()->val($fieldname);
+            }
+            if (isset($this->request->{$valuesSource}[$fieldname])) {
```

Comment:
For unmodified `$_POST` and `$_GET`, yes - for `$this->request->data` and `$this->request->query` not necessarily.

It fails for instance if you do `$this->request->data = $this->Articles->findById($id)` and it does fill some values with `null` which you then overwrite via query sourceValues e.g. `<?= $this->Form->create(null, ['values' => ['data', 'query']]) ?>` - not that I advertise such code.